src/combinatorics/simple_graph/prod.lean

```diff
@@ -0,0 +1,150 @@
+
+/-- Project a walk on `G □ H` to a walk on `G` by discarding the moves in the direction of `H`. -/
+def of_box_prod_left : Π {x y : α × β}, (G □ H).walk x y → G.walk x.1 y.1
+| _ _ nil := nil
+| x z (cons h w) := or.by_cases h (λ hG, w.of_box_prod_left.cons hG.1)
+    (λ hH, show G.walk x.1 z.1, by rw hH.2; exact w.of_box_prod_left)
+
+/-- Project a walk on `G □ H` to a walk on `H` by discarding the moves in the direction of `G`. -/
+def of_box_prod_right : Π {x y : α × β}, (G □ H).walk x y → H.walk x.2 y.2
+| _ _ nil := nil
+| x z (cons h w) := (or.symm h).by_cases (λ hH, w.of_box_prod_right.cons hH.1)
+  (λ hG, show H.walk x.2 z.2, by rw hG.2; exact w.of_box_prod_right)
+
+@[simp] lemma of_box_prod_left_box_prod_left :
+  ∀ {a₁ a₂ : α} (w : G.walk a₁ a₂), (w.box_prod_left H b).of_box_prod_left = w
+| _ _ nil := rfl
+| _ _ (cons' x y z h w) := begin
+  rw [box_prod_left, of_box_prod_left, or.by_cases, dif_pos, of_box_prod_left_box_prod_left],
+  exact ⟨h, rfl⟩,
+end
+
+@[simp] lemma of_box_prod_left_box_prod_right :
+  ∀ {b₁ b₂ : α} (w : G.walk b₁ b₂), (w.box_prod_right G a).of_box_prod_right = w
+| _ _ nil := rfl
+| _ _ (cons' x y z h w) := begin
+  rw [box_prod_right, of_box_prod_right, or.by_cases, dif_pos, of_box_prod_left_box_prod_right],
+  exact ⟨h, rfl⟩,
+end
+
+end walk
+
+variables {G H}
+
+protected lemma preconnected.box_prod (hG : G.preconnected) (hH : H.preconnected) :
+  (G □ H).preconnected :=
+begin
+  rintro x y,
+  obtain ⟨w₁⟩ := hG x.1 y.1,
+  obtain ⟨w₂⟩ := hH x.2 y.2,
+  rw [←@prod.mk.eta _ _ x, ←@prod.mk.eta _ _ y],
+  exact ⟨(w₁.box_prod_left _ _).append (w₂.box_prod_right _ _)⟩,
+end
+
+protected lemma preconnected.of_box_prod_left [nonempty β] (h : (G □ H).preconnected) :
+  G.preconnected :=
+begin
+  classical,
+  rintro a₁ a₂,
+  obtain ⟨w⟩ := h (a₁, classical.arbitrary _) (a₂, classical.arbitrary _),
+  exact ⟨w.of_box_prod_left⟩,
+end
+
+protected lemma preconnected.of_box_prod_right [nonempty α] (h : (G □ H).preconnected) :
+  H.preconnected :=
+begin
+  classical,
+  rintro b₁ b₂,
+  obtain ⟨w⟩ := h (classical.arbitrary _, b₁) (classical.arbitrary _, b₂),
+  exact ⟨w.of_box_prod_right⟩,
+end
+
+protected lemma connected.box_prod (hG : G.connected) (hH : H.connected) : (G □ H).connected :=
+by { haveI := hG.nonempty, haveI := hH.nonempty, exact ⟨hG.preconnected.box_prod hH.preconnected⟩ }
+
+protected lemma connected.of_box_prod_left (h : (G □ H).connected) : G.connected :=
+by { haveI := (nonempty_prod.1 h.nonempty).1, haveI := (nonempty_prod.1 h.nonempty).2,
+  exact ⟨h.preconnected.of_box_prod_left⟩ }
+
+protected lemma connected.of_box_prod_right (h : (G □ H).connected) : H.connected :=
+by { haveI := (nonempty_prod.1 h.nonempty).1, haveI := (nonempty_prod.1 h.nonempty).2,
+  exact ⟨h.preconnected.of_box_prod_right⟩ }
+
+@[simp] lemma box_prod_connected : (G □ H).connected ↔ G.connected ∧ H.connected :=
+⟨λ h, ⟨h.of_box_prod_left, h.of_box_prod_right⟩, λ h, h.1.box_prod h.2⟩
```

Comment:
No, because `empty × V` is empty even if `V` isn't, so `G □ H` can be preconnected with `H` being preconnected. You could say something like
```
@[simp] lemma box_prod_preconnected :
  (G □ H).preconnected ↔ is_empty α ∨ is_empty β ∨ G.connected ∧ H.connected :=
```
but that doesn't seem super-interesting.

Reply:
I meant with `[nonempty α] [nonempty β]` before the colon

Reply:
Sure, but then `G.preconnected ↔ G.connected` and `H.preconnected ↔ H.connected`, so what's the point?